Add weak_excluded packages to the goal (RhBug:1699672)

j-mracek opened the pull request:
It weak_exclude packages for solver and remove them from the list of
candidates for weak deps. It will prevents of installing disfavored
packages that are recommended or supplemented.

Additionally it add mechanism to identify unmet weak deps and
weak_exclude them in Goal.

https://bugzilla.redhat.com/show_bug.cgi?id=1699672

Requires: https://github.com/rpm-software-management/libdnf/pull/1279
Tests: https://github.com/rpm-software-management/ci-dnf-stack/pull/1005

Alternative to: https://github.com/rpm-software-management/dnf/pull/1774

mlschroe commented:
I don't think your supplements handling makes sense. Why would it matter for an installed(!) package if a package it supplements is installed or not? That's the wrong end of the dependency.

I.e. the package *that contains the supplements* is the one that is optional. And as it is currently installed, the user is ok with having it in the system.

j-mracek commented:
> I don't think your supplements handling makes sense. Why would it matter for an installed(!) package if a package it supplements is installed or not? That's the wrong end of the dependency.
> 
> I.e. the package _that contains the supplements_ is the one that is optional. And as it is currently installed, the user is ok with having it in the system.

Good point I will improve it.

j-mracek commented:
I changed the logic for supplemented and moved the code to the libdnf.

The new implementation uses for supplements only available packages not installed in any version. When any of supplements is already installed, the packages is weak_excluded. It means that any additional packages with supplements will be not installed after supplemented package is installed. It is really not perfect, but without storing additional information about available supplementing packages at the time of installation or other operations (downgrade, upgrade, ...) we cannot do it better. In case when we will store all information required for better resolvement we will collect probably to much information during a longer time period.

j-mracek commented:
With auto detection of unmet supplements I see 3 possibilities:
1. Use detection described in https://github.com/rpm-software-management/dnf/pull/1780#issuecomment-869462452
2. Collect additional data and make a decision based on full history
3. Do not use any auto-detection and keep only manual adjustment using the configuration option `weak_excludes`
Please I have an interest to hear any opinion or other options.

adelton commented:
The `excluded` and `not excluded` messages are likely useful for debugging but they might be distracting to people who do not want to dive into the details of the operation.

pkratoch commented on dnf/base.py:
```diff
@@ -838,6 +853,9 @@ def resolve(self, allow_erasing=False):
         goal.add_protected(self.sack.query().filterm(
             name=self.conf.protected_packages))
+
+        self._set_excludes_from_weak_to_goal()
```
Could the new function in the context part of libdnf be used instead of duplicating it in dnf?

j-mracek replied:
It is good suggestion but it will be hard but not impossible, because the configuration class in python is quite difficult to pass to c code (The configuration class (C++) is exported by SWIG, then repacked by python in Class. Goal class (C++) is available in Python be C-Python). In DNF5 it must be not duplicated.

pkratoch commented on doc/conf_ref.rst:
```diff
@@ -525,7 +525,21 @@ configuration file by your distribution to override the DNF defaults.
     in Configuration reference.
 
-.. _conf_repo_options-label:
+.. _exclude_from_weak-label:
```
The other options are sorted alphabetically. Can you please move these to the correct place?

j-mracek replied:
Thanks, the problem was created by rename of options.

pkratoch commented on doc/conf_ref.rst:
```diff
@@ -525,7 +525,21 @@ configuration file by your distribution to override the DNF defaults.
+    :ref:`list <list-label>`
+
+    Prevent installing packages specified by a name or a glob and separated by a comma to satisfy recommends
+    or supplements dependency. Defaults to ``[]``.
```
I'd separate the description of how to specify the packages from the rest to improve the readability.

My suggestion: Prevent installing packages as weak dependencies (recommends or supplements). The packages can be specified by a name or a glob and separated by a comma. Defaults to ``[]``.

j-mracek replied:
Thanks, I used your suggested version.

pkratoch commented on doc/conf_ref.rst:
```diff
@@ -525,7 +525,21 @@ configuration file by your distribution to override the DNF defaults.
+    :ref:`boolean <boolean-label>`
+
+    If enabled it will autodetect unmet weak dependencies (recommends or supplements) of installed packages and block
+    installation of packages satisfying unmet dependencies. Defaults to ``true``.
```
I have couple of notes to this one, since it's very hard to find a description of this option that would be accurate yet easily understandable.

Firstly, my suggestion: If enabled, dnf will autodetect unmet weak dependencies (recommends or supplements) of already installed packages and won't install providers of these dependencies as weak dependencies (but they can still be installed to satisfy regular dependencies). Defaults to ``true``.

Explanation of the changes I suggested:
1. Minor improvements: add comma after "if enabled" and replace "it" for "dnf".
2. Add "already" before "installed packages" to make it immediately clear that we are talking about packages that are already on the system and not packages that are currently being installed.
3. Replace "block installation of" with "won't install", because "block" might seem unnecessarily radical (like the whole transaction will be blocked or something).
4. The "packages satisfying unmet dependencies" seems very confusing to me. I'm not sure if my suggestion is good enough. Other alternative might be to replace the ending with: "won't install packages that satisfy these autodetected weak dependencies as weak dependencies"
5. Additional explanatory sentence into parentheses afterwards just to make it really clear.

Let me know if this seems ok to you. Ideally, I'd like to then ask someone else to read the description, because I'm biased at this point (and my English is not that good).

j-mracek replied:
Thanks, I used your suggested version with a minor change. I skipped "already" from point 2. Somehow I feel that it is redundant.

lukash commented on doc/conf_ref.rst:
```diff
@@ -208,6 +208,23 @@ configuration file by your distribution to override the DNF defaults.
+
+    If enabled, dnf will autodetect unmet weak dependencies (recommends or supplements) of installed packages and won't
+    install providers of these dependencies as weak dependencies (but they can still be installed to satisfy regular
+    dependencies). Defaults to ``true``.
```
Does this only affect upgrade/downgrade/reinstall operations or does it also mean that, having package A installed on the system that has a weak dep W which is not installed, if I install package B which also has a weak dep W, W will not be installed?

I think this should be clarified. Otherwise the text is very precise, maybe not easy to understand at first, after the above detail is added, maybe a short "in other words" could be added too to help with understanding what it does.

j-mracek replied:
It affects all operations

pkratoch replied:
I agree we should clarify that unrelated package B also won't bring in W, since this is a serious limitation of the autodetection. So, how about this:

If enabled, dnf will autodetect unmet weak dependencies (recommends or supplements) of installed packages and won't install providers of these dependencies as weak dependencies (even if they satisfy weak dependency of another package). However, they can still be installed to satisfy regular dependencies. Defaults to ``true``.

j-mracek replied:
Fixed

lukash replied:
> If enabled, dnf will autodetect unmet weak dependencies (recommends or supplements) of installed packages and won't install providers of these dependencies as weak dependencies (even if they satisfy weak dependency of another package). However, they can still be installed to satisfy regular dependencies. Defaults to true.

@pkratoch while I guess it technically answers my question, it's really not obvious until you think hard about it. What about:
> If enabled, dnf will autodetect unmet weak dependencies (recommends or supplements) of packages installed on the system. Providers of these weak dependencies will not be installed by dnf as weak dependencies any more (they will still be installed if pulled in as a regular dependency).

j-mracek replied:
Thank you both for improvements

j-mracek replied:
I pushed an update
